awscli/customizations/history/db.py

```diff
@@ -0,0 +1,256 @@
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import os
+import uuid
+import time
+import json
+import base64
+import datetime
+import threading
+import logging
+from collections import MutableMapping
+
+from botocore.history import BaseHistoryHandler
+
+from awscli.compat import sqlite3
+from awscli.compat import ensure_text_type
+from awscli.compat import binary_type
+
+
+LOG = logging.getLogger('awscli.history.db')
+
+
+class DatabaseConnection(object):
+    _CREATE_TABLE = """
+    CREATE TABLE IF NOT EXISTS records (
+      id TEXT,
+      request_id TEXT,
+      source TEXT,
+      event_type TEXT,
+      timestamp INTEGER,
+      payload TEXT
+    )"""
+    _ENABLE_WAL = """PRAGMA journal_mode=WAL"""
+    _DATABASE_NAME = 'history.db'
+    _DEFAULT_DATABASE_FILENAME = os.path.expanduser(
+        os.path.join('~', '.aws', 'cli', 'history', _DATABASE_NAME))
+
+    def __init__(self, db_filename=None):
+        if db_filename:
+            self._db_filename = db_filename
+        elif os.environ.get('AWS_HISTORY_PATH'):
+            self._db_filename = os.path.join(
+                os.environ.get('AWS_HISTORY_PATH'), self._DATABASE_NAME)
+        else:
+            self._db_filename = self._DEFAULT_DATABASE_FILENAME
+
+        if not os.path.isdir(os.path.dirname(self._db_filename)):
+            os.makedirs(os.path.dirname(self._db_filename))
+        self._connection = sqlite3.connect(
+            self._db_filename, check_same_thread=False)
+        self._ensure_database_setup()
+
+    def _ensure_database_setup(self):
+        self._create_record_table()
+        self._try_to_enable_wal()
+
+    def _create_record_table(self):
+        cursor = self.cursor()
+        cursor.execute(self._CREATE_TABLE)
+        self._connection.commit()
+
+    def _try_to_enable_wal(self):
+        try:
+            cursor = self.cursor()
+            cursor.execute(self._ENABLE_WAL)
+            self._connection.commit()
+        except sqlite3.Error:
+            # This is just a performance enhancement so it is optional. Not all
+            # systems will have a sqlite compiled with the WAL enabled.
+            pass
+
+    def cursor(self):
+        return self._connection.cursor()
+
+    def commit(self):
+        self._connection.commit()
+
+    @property
+    def row_factory(self):
+        return self._connection.row_factory
+
+    @row_factory.setter
+    def row_factory(self, row_factory):
+        self._connection.row_factory = row_factory
+
+
+class PayloadSerializer(json.JSONEncoder):
+    def _encode_mutable_mapping(self, obj):
+        return dict(obj)
+
+    def _encode_binary(self, obj):
```

Comment:
This seems hacky.  What were the specific history records that resulted in needing this method?